### PR TITLE
Batch & Transaction Memory Storage

### PR DESCRIPTION
**Pull request by jacque006**
Resolves https://github.com/thehubbleproject/hubble-contracts/issues/548

Out of scope:
 - Marking batches/txns as finalized.

### TODO
- [x] Link/hookup existing functionality to current implementations in client.
- [x] Unit/integration tests.
- [x] Add doc comments around all new interfaces/types/implementations.
~- [ ] (optional?) Integrate finalization state transitions~ https://github.com/thehubbleproject/hubble-contracts/issues/592

### CODE REVIEW

File: test/client/storage.test.ts [hunk 1]
```diff
@@ -100,16 +103,12 @@ describe("StateMemoryEngine", () => {
...
-            } catch (err) {
-                errMsg = err.message;
-            }
-            assert.equal(
-                errMsg,
-                `Tree at level ${depth -
-                    maxSubtreeDepth} is full, no room for subtree insert`
+            assert.isRejected(
+                smallTreeEngine.findVacantSubtree(maxSubtreeDepth),
+                new RegExp(
```
Comment (ChihChengLiang) [inline test/client/storage.test.ts:108]: The regex match looks fragile to the change of the error message.
What do you think about creating a custom error class `TreeAtLevelIsFull` like what we did in the `exceptions.ts`, so we can check against `error.name` and `error.level`?

File: ts/client/storageEngine/transactions/interfaces.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { Batch, OffchainTx } from "../../features/interface";
+import { Status } from "./constants";
+
+export type TransactionStatus = Readonly<{
+    status: Status;
+    detail?: string;
+    batchID?: number;
+    blockIncluded?: number;
+    transaction: OffchainTx;
+    batch?: Batch;
```
Comment (jacque006) [inline ts/client/storageEngine/transactions/interfaces.ts:10]: Include L1 transaction L2 txn is part of

File: ts/client/storageEngine/transactions/constants.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+export enum Status {
+    // Transaction has been submitted to proposer
+    Pending = "pending",
+    // Transaction has been submitted as part of a batch
+    Submitted = "submitted",
+    // Transaction was finalized
+    Finalized = "finalized",
+    // Transaction was rolled back
+    RolledBack = "rolledback",
```
Comment (jacque006) [inline ts/client/storageEngine/transactions/constants.ts:9]: Move to new issue covering transaction state after a rollback, remove from here.

Comment (jacque006) [reply ts/client/storageEngine/transactions/constants.ts:9]: https://github.com/thehubbleproject/hubble-contracts/issues/585

File: ts/client/services/rpc.ts [hunk 4]
```diff
@@ -25,31 +25,46 @@ export class RPC {
...
+                }
+            );
+            fastify.post<{ Body: { bytes: string } }>(
+                "/tx",
+                tx,
+                async request => {
+                    const bytes = arrayify(request.body.bytes);
+                    const transfer = TransferOffchainTx.deserialize(bytes);
+                    console.log(transfer.toString());
+                    depositPool.pushDeposit(transfer.message());
```
Comment (ChihChengLiang) [inline ts/client/services/rpc.ts:51]: Should this be the transferPool instead of depositPool?

Comment (jacque006) [reply ts/client/services/rpc.ts:51]: Yes, my mistake. Will correct.

**Comment by jacque006** [comment]
@ChihChengLiang Ready for your review.

**Comment by jacque006** [comment]
One other thing I should note: A lot of the work in this PR is tightly coupled with different related domain concepts (batches, transactions) having to know explicitly about each other in order to properly store their state. As we build more functionality out, it might make sense to use some sort of message bus like a NodeJS `EventEmitter` to allow these different domains to talk to each other via eventing without explicitly calling cross domain in code. This would also reduce the services/managers (dependencies) that any given piece would need to function.

I experimented with an async event emitter in one iteration of this but ran into some issues. I will likely revisit this concept in future work when and where it makes sense.

File: ts/client/storageEngine/transactions/interfaces.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+     * Transitions a pending transaction to being submitted
+     * in an L1 txn (batch submission).
+     *
+     * @param message Message or object of the L2 transaction.
+     * @param meta Metadata about the submission, including batch and L1 txn info.
+     */
+    submit(
+        msgOrTxn: TransationMessageOrObject,
+        meta: SubmitMeta
+    ): Promise<TransactionStatus>;
```
Comment (jacque006) [inline ts/client/storageEngine/transactions/interfaces.ts:65]: `submitted`, use past tense verbs

File: ts/client/features/transfer.ts [hunk 6]
```diff
@@ -93,9 +95,26 @@ export class TransferCompressedTx implements CompressedTx {
             ]
         );
     }
+
+    public toString(): string {
+        return `<TransferCompressed ${this.fromIndex}->${this.toIndex} $${this.amount}  fee ${this.fee}>`;
```
Comment (jacque006) [inline ts/client/features/transfer.ts:100]: Extra space

File: ts/client/services/events/newBatch.ts [hunk 7]
```diff
@@ -46,8 +52,25 @@ export class NewBatchEventSyncer extends ContractEventSyncer {
...
+
+        const txns = await this.batchHandlingContext.processBatch(batch);
+        for (const tx of txns) {
+            // TODO Figure out finalization
+            // https://github.com/thehubbleproject/hubble-contracts/issues/592
+            const finalized = false;
+            const l1BlockIncluded = -1;
+
+            await this.transactionStorage.sync(tx, {
+                finalized,
```
Comment (jacque006) [inline ts/client/services/events/newBatch.ts:67]: This can likely be removed.

File: ts/exceptions.ts [hunk 8]
```diff
@@ -157,3 +173,35 @@ export class ZeroAmount extends StateTreeExceptions {
...
+        this.name = "TransactionAlreadyExists";
+    }
+}
+
+export class StatusTransitionInvalid extends TransactionStatusException {
+    constructor(cur: Status, next: Status) {
+        super(`cannot transition from status ${cur} to ${next}`);
+        this.name = "StatusTransitionInvalid";
+    }
+}
```
Comment (jacque006) [inline ts/exceptions.ts:207]: Use `HubbleError`

**Review by ChihChengLiang — APPROVED** [review]
Looks good :shipit:!

**Pull request merged**